### DISCUSSION TITLE
Can we Automate IP Address Management and Device Provisioning with help of NetBox ?

### DISCUSSION THREAD

**Discussion by Jaz108** [netbox-community/netbox#14623] [Help Wanted!] [answered]
Hi Guys I am new to Netbox, and want to ask a question. 

**Introduction:**
I am Devops Engineer in my org, My Infra is scattered over various platforms, like GCP , Vshpere and Some Local Instances.
Currently theres no IPAM tool we are using, we are thinking to use Netbox for this Purpose. We want to Automate IPAM, A centralized Source of truth for all the machines and IPs.

**Specific Questions:**

1. Is it possible with some integration that , If we create some new instaces in GCP or Vshpere and it gets listed on Netbox automatically ?
2. Will NetBox be the right tool to achieve this goal ? 

Thanks for any help.

**Comment by candlerb** [comment, marked as answer]
Yes you can do this, but you'll have to do the integration work yourself.

You'll also have to decide which way the flow goes. Options include:
1. You create the VM first in GCP or vSphere, and the assignment gets recorded in Netbox. In this case, Netbox isn't really acting as a "source of truth" but just a record of after-the-fact changes.
2. You create the VM first in Netbox, and this triggers creation of the VM in GCP or vSphere by pushing to their respective APIs. This is the full "source of truth" active provisioning flow, but the Netbox data model will need to include all the data required to supply the API with necessary provisioning fields: you may need to add some Custom Fields for this.
3. Hybrid: you reserve the IP address first in Netbox, then create the VM in GCP or vSphere manually using the assigned IP, and then create the VM model in Netbox. This makes Netbox just the IPAM source of truth for assigning unique IPs.

Whether this is the "right" tool for you is a matter of what you're trying to achieve.

If you only want to reflect the current live state of your network - "what VMs are running right now?" - and you can include an agent as part of your provisioning flow, then you might be better off using a monitoring system and dashboard (e.g. prometheus+grafana, check_mk, librenms etc). This would be especially true if your environment is highly dynamic, e.g. VMs created/removed using auto scaling groups, or Kubernetes containers.

**Reply by Jaz108** [reply]
Hi @candlerb, Thanks for the reply, we are trying to achieve what you said in Your first point, But do we need to record the IP manually ? I want a integration such that, We achieve it automatically or periodically triggering a Function .

**Reply by candlerb** [reply]
No: you can use the REST API to assign "next available IP address" from a prefix: either just to pick it (GET), or to assign and create it atomically (POST).

But if you want to model the whole Virtual Machine then it needs to be more than an IP address: it's at least three objects (VM + interface + IP address). You can automate the creation of those too.

There's also a certain amount you can do within Netbox itself. For example, here's a [Custom Script](https://docs.netbox.dev/en/stable/customization/custom-scripts/) which creates a (VM + interface + IP address) combo, although it requests the IP address as a manual field rather than allocating it:
https://github.com/netbox-community/customizations/blob/master/scripts/create_vm.py

**Reply by Jaz108** [reply]
Ohh Okay, thanks I have gathered various informations today, Such that I am not using Netbox as source of truth, and Few scripts to populate netbox, There is another script called netbox-sync, currently i dont have link to it, But i will update it once I have it .
Thanks for the help, I hope i can ask more questions on this thread. Once again, thank you very much
